### ISSUE TITLE
Can't assign array elements to frozen strings using `start` and `length` parameters

### ISSUE THREAD

**Issue by deivid-rodriguez** [oracle/truffleruby#2355]
When using the `array[start, length] = object → object` form of `Array#[]=`, truffleruby crashes when assigning a frozen string literal while CRuby works just fine:

##### On CRuby

```
$ ruby -e 'a = ["1","2","3"]; a[0,2]="4".freeze; puts a'
4
3
```

##### On Truffleruby

```
$ ruby -e 'h = ["1","2","3"]; h[0,2]="4".freeze; puts h'
-e:1:in `[]=': TruffleRuby doesn't have a case for the org.truffleruby.core.cast.ArrayCastNodeGen node with values of type org.truffleruby.core.string.ImmutableRubyString (TypeError)
	from org.truffleruby.core.cast.ArrayCastNodeGen.executeAndSpecialize(ArrayCastNodeGen.java:206)
	from org.truffleruby.core.cast.ArrayCastNodeGen.execute(ArrayCastNodeGen.java:83)
	from org.truffleruby.core.array.ArrayConvertNode.execute(ArrayConvertNode.java:29)
	from org.truffleruby.core.array.ArrayNodes$SetIndexNode.setTernary(ArrayNodes.java:429)
	from org.truffleruby.core.array.ArrayNodesFactory$SetIndexNodeFactory$SetIndexNodeGen.executeAndSpecialize(ArrayNodesFactory.java:1466)
	from org.truffleruby.core.array.ArrayNodesFactory$SetIndexNodeFactory$SetIndexNodeGen.execute(ArrayNodesFactory.java:1375)
	from org.truffleruby.language.RubyCoreMethodRootNode.execute(RubyCoreMethodRootNode.java:52)
	from org.graalvm.compiler.truffle.runtime.OptimizedCallTarget.executeRootNode(OptimizedCallTarget.java:617)
	from -e:1:in `<main>'
```

**Comment by eregon** [comment]
Thank you for the report, ArrayCastNode was incorrect, I'll fix it.

**Comment by eregon** [comment]
Fixed now, will be in 21.2, and already in nightly builds.

**Comment by deivid-rodriguez** [comment]
Thanks!
